[CodeGen] Generalize trap emission after SP check fail

duk-37 opened the pull request:
Generalize and improve some target-specific code that emits traps after stack protector failure in SelectionDAG & GlobalIsel.

arsenm commented on llvm/test/CodeGen/AArch64/stack-protector-trap-unreachable.ll:
```diff
@@ -0,0 +1,34 @@
+; NOTE: Do not autogenerate, we'll lose the important NOT- check otherwise.
+;; Make sure we emit trap instructions after stack protector checks iff NoTrapAfterNoReturn is false.
```
-NEXT checks that exclude the instruction are superior to -NOT checks, which will not fail if the output ever changes

duk-37 replied:
Noted - I'll add some -NEXT checks too. I'd like to keep the -NOT check since it makes it explicit what we're testing for.

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -3807,16 +3807,11 @@ bool IRTranslator::emitSPDescriptorFailure(StackProtectorDescriptor &SPD,
-  // because the function return type can be different from __stack_chk_fail's
-  // return type (void).
-  const TargetMachine &TM = MF->getTarget();
-  if (TM.getTargetTriple().isPS() || TM.getTargetTriple().isWasm()) {
```
This is loosing the target specific behaviors? TargetOptions aren't really options for the target. No test changes for this?

duk-37 replied:
There are existing tests for these targets

duk-37 committed:
[CodeGen] Generalize trap instruction emission after SP check fail

Generalize some target-specific code that emits traps after stack protector failure in SelectionDAG/GlobalIsel and add some tests.

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-stack-protector-trap-unreachable.ll:
```diff
@@ -0,0 +1,28 @@
+
+entry:
+  %buf = alloca [8 x i8]
+  %2 = call i32(ptr) @callee(ptr %buf) nounwind
```
```suggestion
  %result = call i32(ptr) @callee(ptr %buf) nounwind
```

arsenm commented on llvm/test/CodeGen/AArch64/stack-protector-trap-unreachable.ll:
```diff
@@ -0,0 +1,36 @@
+
+entry:
+  %buf = alloca [8 x i8]
+  %2 = call i32(ptr) @callee(ptr %buf) nounwind
```
```suggestion
  %result = call i32(ptr) @callee(ptr %buf) nounwind
```

arsenm commented on llvm/test/CodeGen/AArch64/stack-protector-trap-unreachable.ll:
```diff
@@ -0,0 +1,36 @@
+; RUN: llc -mtriple=aarch64 -fast-isel -verify-machineinstrs \
+; RUN:     -trap-unreachable -no-trap-after-noreturn=false -o - %s | FileCheck --check-prefix=TRAP_UNREACHABLE %s
+
+define void @test() nounwind ssp {
```
Is it possible to run the test on the post-SSP pass IR?

duk-37 replied:
Yes - and this lets us autogenerate the GlobalIsel test.

arsenm reviewed (commented):
essentially lgtm

duk-37 commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -3183,15 +3183,10 @@ SelectionDAGBuilder::visitSPDescriptorFailure(StackProtectorDescriptor &SPD) {
-  // Passing 'true' for doesNotReturn above won't generate the trap for us.
-  if (TM.getTargetTriple().isPS())
-    Chain = DAG.getNode(ISD::TRAP, getCurSDLoc(), MVT::Other, Chain);
-  // WebAssembly needs an unreachable instruction after a non-returning call,
-  // because the function return type can be different from __stack_chk_fail's
-  // return type (void).
```
There are already some comments in the target code, but I've also added the relevant `__stack_chk_fail` notes.
